File: src/content/3.7/code/scala/snippet20.scala
implicit val streamComonad = new Comonad[Stream] {
  def extract[A](wa: Stream[A]): A =
    wa match {
      case Stream(a, _) => a()
    }

  def duplicate[A](wa: Stream[A]): Stream[Stream[A]] = wa match {
    case s @ Stream(a, as) =>
      Stream(() => s, () => duplicate(as()))
  }
}

File: src/content/3.7/code/ocaml/snippet20.ml
(* Implement Extract *)
module StreamComonadBase (F : Functor with type 'a t = 'a stream) :
  ComonadBase with type 'a w = 'a stream = struct
  type 'a w = 'a stream

  include F

  let extract (Cons (x, _)) = x
end

(* Implement duplicate *)
module StreamComonadDuplicate :
  ComonadDuplicate with type 'a w = 'a stream = struct
  type 'a w = 'a stream

  let rec duplicate (Cons (x, xs)) =
    Cons (Cons (x, xs), Lazy.from_val (duplicate (Lazy.force xs)))
  ;;
end

(* Generate full Comonad Instance *)
module StreamComonad : Comonad with type 'a w = 'a stream =
  ComonadImplViaExtend
    (StreamComonadBase (StreamFunctor)) (StreamComonadDuplicate)
